### DISCUSSION TITLE
Why store `.keepme` in `bin/`?

### DISCUSSION THREAD

**Discussion by abitrolly** [Homebrew/discussions#124] [Everyday usage]
```shell
~ ls -lA .linuxbrew/bin
total 0
lrwxrwxrwx. 1 anatoli anatoli 42 Nov  8 14:27 brew -> /home/anatoli/.linuxbrew/Homebrew/bin/brew
lrwxrwxrwx. 1 anatoli anatoli 28 Nov  9 07:01 k9s -> ../Cellar/k9s/0.23.9/bin/k9s
-rw-r--r--. 1 anatoli anatoli  0 Nov  9 07:00 .keepme
lrwxrwxrwx. 1 anatoli anatoli 36 Nov  9 07:01 patchelf -> ../Cellar/patchelf/0.11/bin/patchelf
```
I don't see why it is needed on Linux. There is a similar question https://github.com/Homebrew/brew/issues/6263#issuecomment-533218217 by @KizzyCode and answers with vague references to shady OS X engineers.

**Comment by jonchang** [comment]
If this is a problem for you, please consider opening a pull request to fix it.

**Reply by abitrolly** [reply]
The problem is that there is not enough info.

**Reply by SMillerDev** [reply]
It seems to be needed on macOS and it doesn't do anything on linux, I'm not sure what other info you need

**Reply by abitrolly** [reply]
The exact info I need is why it is needed it macOS? )

**Reply by Bo98** [reply]
It's needed on macOS to prevent the directories being removed on OS upgrades. `brew` requires the directories to exist, even if empty.

**Reply by gromgit** [reply]
https://github.com/Homebrew/brew/issues/6263

**Reply by abitrolly** [reply]
@Bo98 do you mean that on upgrades macOS erases anything in `/usr/local/` it doesn't know about, as well as everything dot directory in in `$HOME`?

**Reply by abitrolly** [reply]
@gromgit I mentioned this issue already in original message. What's the point?

**Reply by gromgit** [reply]
Sorry, cross-eyed from multiple migrations, didn't notice the link in your OP.

My point is that you're probably better off asking the Apple engineers directly, if you absolutely positively must know.

**Reply by Bo98** [reply]
The macOS installer removes empty directories (or at least used to) in `/usr/local` and it requires `sudo` permissions to recreate them. The `.keepme` file was placed to prevent the directories from being removed and thus not need `sudo` to recreate them when they are later filled.

`brew` doesn't have a dot directory in `$HOME` on macOS. `/usr/local` on macOS == `/home/linuxbrew/.linuxbrew` on Linux.

**Comment by abitrolly** [comment]
@Bo98 is that `macOS installer` software from Apple? I expect installers to remove empty directories on uninstall, not just when invoked for some other program. When the installer starts to remove these dirs if not during uninstall of `brew` itself?

**Reply by jonchang** [reply]
The pull request you linked refers to the Catalina installer, so this is when upgrading macOS versions.

**Reply by abitrolly** [reply]
@jonchang thanks for the clarification. It could have have been a `brew` installer for Catalina. So does it clean up everything unknown in `/usr/local` and dot dirs in `$HOME`?

**Reply by SMillerDev** [reply]
It isn't a brew installer since we don't have installers for specific macOS versions. The macOS installer cleans, or at least used to, unknown empty directories in `/usr/local`. That's why we put a `.keepme` file there as instructed by Apple engineers (as mentioned in the issue you linked)

**Reply by abitrolly** [reply]
@SMillerDev but if the `brew` is installed, it is located in `bin/` directory, right?  So it can not be empty.
